design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,160 @@
+* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.
+
+Cons:
+* We’re adding yet another semi alpha knob to an already complicated set of knobs.
+
+# Fix Alternatives
+## Alternative #1 Adding `full-pcpus-only` option as a global configuration
+As opposed to the suggested solution where the user needs to add this annotation to every VM, it is possible to make this behavior global. The assumption is that clusters run homogeneous Nodes, so this will eliminate the need to add this annotation every time.
+
+Pros:
+* VM configuration is less complex.
+* Similar to the way default runtimeclass is configured on Kubevirt.
+
+Cons:
+* The suggested annotation is waisting a CPU when used on nodes with hyperthreading disabled. A cluster-wide configuration may be wasteful for heterogeneous clusters, where node-level granularity is needed.
+
+## Alternative #2 Configure both pod and Guest
+It is possible to use both the current Guest API and the old yet existing pod API, like so:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        resources:
+          requests:
+            cpu: '10'
+          limits:
+            cpu: '10'
+        cpu:
+          cores: 4
+          sockets: 1
+          threads: 2
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+```
+In order to do that the current validating webhooks regarding dedicatedCpuPlacement and IsolatedEmulatorThread will have to be loosened.
+
+Pros:
+* This option does not introduce any alpha knobs.
+
+Cons:
+* This option exposes the user to unnecessary implementation details and possible waste of physical CPUs. They have to do things manually, opening the door for misuse. Loosening the validating webhook doesn't make sense in the general case.
+* KubeVirt will need to adjust internal assumptions that are based on the fact that VMs {req, limits}.cpu = vCPUs. Specifically, how the housekeeping cgroup is being formed, but possibly other aspects.
+* VMs that were defined like this will have to change when the long term solution will arrive (backwards-compatibility issue)
+
+## Scalability
+The solution proposed is “wasting” one unused CPU for the purpose of maintaining SMT alignment. I believe it is a minor waste considering the alternatives, such as disabling SMT on the node. As I consider the solution adequately scalable.
+
+## Update/Rollback Compatibility
+Since the annotation is not an API change, it is backportable.
+When a better solution is possible, Kubevirt can simply replace the annotation support with a better approach, thus not making the change relatively seamless for the client. 
+
+## Functional Testing Approach
+Will add unit tests to `renderresources_test.go` to check extra housekeeping logic, and in `templates.go` to check the virt-launcher pod spec output.
```

Comment:
Does an e2e test is also needed?

Reply:
Good question. 
I think we will encounter a problem with this, since there is no support for SMT on kubevirtci, nor is there a convenient way to set the cpu-manager Policy there. 
Simply put - we can't replicate the SMT alignment error. Hence, I would only add unit tests.
@vladikr what do you think?

Reply:
The main issue is not the SMT support on the nodes, but rather the k8s CPU manager beta policy configuration. I'm not sure how we can have such a configuration in kubevirt CI without creating a dedicated lane.
@xpivarc what do you think?

Reply:
@vladikr  Don't we need both "hardware" and Kubelet configuration?

Reply:
@xpivarc the smt can be adjusted. I imagine that we can do the same trick we did before. We can set one of the nodes to enable smt.
Theoretically, we could change the kubelet configuration, but I fear that it will be tricky since the dynamic configuration has been deprecated.

Reply:
We could but I am not sure about the ROI. Maybe unit tests are enough here?

Reply:
+1 for unit tests only unless e2e is absolutely required.

Reply:
IMO all the changes done on the first phase PR are covered by unit tests, making the e2e non-required. 
Regarding the second phase (cgroups PR WIP) - that part is less covered by unit tests so I'll try to refactor it so that we don't need any e2e tests.

Reply:
I don't mind if we will have unit tests only with this feature.
In the general case, unit tests are not enough. Most of the code is never exercised together so there is no way to verify the whole flow.

This looks more as an esoteric use case that relies on a feature-gated option of the CPU manager that is not likely to be widely used upstream and has no plans to get scheduler support.

Reply:
Ok, I think we have an agreement for this one.